### PR TITLE
Add naive backtracking parser

### PR DESCRIPTION
**Pull request by sahejsingh**
Initial Implementation of a Backtracking Parser.

@anz-bank/sysl-developers

### CODE REVIEW

File: sysl2/parser.go [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+	s1 := makeSequence(b, a, b)
+	s2 := makeSequence(b, A)
+	choiceS := sysl.Choice{Sequence: []*sysl.Sequence{s1, s2}}
+	ruleS := sysl.Rule{Name: ruleNameS, Choices: &choiceS}
+
+	prods := map[string]*sysl.Rule{}
+	prods["S"] = &ruleS
+	prods["A"] = &ruleA
+
+	return sysl.Grammar{Name: "test", Rules: prods}
```
Comment (marcelocantos) [inline sysl2/parser.go]: Make this function one big literal. Ditto for every make… function below.

File: sysl2/sysl_parser.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+func parse(g sysl.Grammar, input []int, val interface{}) (bool, []int, []interface{}) {
+	if len(input) == 0 {
+		fmt.Println("got input length zero!!!")
+	}
+	result := false
+	tree := make([]interface{}, 0)
+
+	switch val.(type) {
+	case *sysl.Sequence:
+		{
```
Comment (marcelocantos) [inline sysl2/sysl_parser.go]: `case`-level block scope is redundant.

File: sysl2/sysl_parser.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+				matchCount := 0
+				res := false
+				subTree := make([]interface{}, 0)
+
+				for matchCount < maxCount {
+					var remaining []int
+					var matchedTerm interface{}
+					switch t.Atom.Union.(type) {
+					case *sysl.Atom_Choices:
+						{
```
Comment (marcelocantos) [inline sysl2/sysl_parser.go]: ditto

File: sysl2/sysl_parser.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+								remaining = input[1:]
+								matchedTerm = t
+							}
+						}
+					}
+					if res {
+						matchCount++
+						input = remaining
+						subTree = append(subTree, matchedTerm)
+					} else {
```
Comment (marcelocantos) [inline sysl2/sysl_parser.go]: Unnest the inner `if`.

File: sysl2/sysl_parser.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+			tree = append(tree, nil)
+		}
+	case *sysl.Choice:
+		{
+			p := val.(*sysl.Choice)
+			fmt.Printf("choices count : (%d)\n", len(p.Sequence))
+			for i, alt := range p.Sequence {
+				fmt.Printf("trying choice (%d)\n", i)
+				res, remaining, subTree := parse(g, input, alt)
+				if res == true {
```
Comment (marcelocantos) [inline sysl2/sysl_parser.go]: Remove `== true`.

File: sysl2/sysl_parser.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+			fmt.Println(tree)
+			return result, input, tree
+
+		}
+	case *sysl.Rule:
+		{
+			r := val.(*sysl.Rule)
+			fmt.Println("got " + r.GetName().Name)
+			res, remaining, subTree := parse(g, input, r.Choices)
+			if res == true {
```
Comment (marcelocantos) [inline sysl2/sysl_parser.go]: Remove `== true`.

File: sysl2/sysl_parser.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+			p := val.(*sysl.Choice)
+			fmt.Printf("choices count : (%d)\n", len(p.Sequence))
+			for i, alt := range p.Sequence {
+				fmt.Printf("trying choice (%d)\n", i)
+				res, remaining, subTree := parse(g, input, alt)
+				if res == true {
+					fmt.Printf("matched choice :(%d)\n", i)
+					choice := make(map[int][]interface{})
+					choice[i] = subTree
+					tree = append(tree, choice)
```
Comment (marcelocantos) [inline sysl2/sysl_parser.go]: Use map literal inside `append(…)`.

File: sysl2/sysl_parser.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+	return result, input, tree
+}
+
+// CheckGrammar returns true if the grammar consumes the whole string
+// g - grammar to use
+// text - to parse
+// Start rule
+func checkGrammar(g sysl.Grammar, tokens []int, start string) bool {
+
```
Comment (marcelocantos) [inline sysl2/sysl_parser.go]: Redundant empty line.

File: sysl2/sysl_parser.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+				} else if len(subTree) == 1 {
+					tree = append(tree, subTree[0])
+				}
+			}
+
+			fmt.Println("out of loop")
+			result = true
+			fmt.Println(tree)
+			return result, input, tree
+
```
Comment (marcelocantos) [inline sysl2/sysl_parser.go]: Redundant empty line

File: sysl2/parser.go [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+	prods["lhs"] = &lhs
+	prods["rhs"] = &rhs
+	prods["choice"] = &choice
+	prods["seq"] = &seq
+	prods["term"] = &term
+	prods["atom"] = &atom
+	prods["quantifier"] = &quantifier
+
+	return sysl.Grammar{Name: "EBNF", Rules: prods}
+
```
Comment (marcelocantos) [inline sysl2/parser.go]: Redundant empty line. Ditto below.

File: sysl2/Lexer_test.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"testing"
+)
+
+func TestLexer(t *testing.T) {
+	content := `keywords: 
+				while | return | 
+				if | else | import`
+
+	regexs := make([]string, 7)
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/Lexer_test.go]: the index in regexes is used in tokens array. It is easier to read this way.

File: sysl2/Lexer_test.go [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+	regexs[4] = `[?]`
+	regexs[5] = `import`
+	regexs[6] = `[a-z][a-z0-9]*`
+
+	tokens := []int{6, 1, 6, 0, 6, 0, 6, 0, 6, 0, 5, -1}
+
+	lex := makeLexer(content, regexs)
+	for i := range tokens {
+		tok := lex.nextToken()
+		if tok != tokens[i] {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/Lexer_test.go]: done.

File: sysl2/Lexer_test.go [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+			t.Errorf("got the wrong token: %d", i)
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/Lexer_test.go]: changed the Errorf call.
Plan to define Token struct that will carry the actual text that was matched.

File: sysl2/lexer.go [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+	"sort"
+
+	"anz-bank/sysl/sysl2/proto"
+)
+
+type lexer struct {
+	currentIndex     int
+	regexs           []*regexp.Regexp
+	content          string
+	WS               *regexp.Regexp
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/lexer.go]: done.

File: sysl2/lexer.go [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+	for _, s := range choice.Sequence {
+
+		for _, t := range s.Term {
+			if t == nil {
+				continue
+			}
+			var str string
+			a := t.Atom
+			switch a.Union.(type) {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/lexer.go]: yes. I think i have used the same construct in other places.

File: sysl2/lexer.go [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+
+	return arr
+}
+
+// Regular whitespace delimited
+func makeLexer(content string, regexs []string) lexer {
+
+	compiledRegexes := make([]*regexp.Regexp, len(regexs))
+	for i, reg := range regexs {
+		compiledRegex, err := regexp.Compile(reg)
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/lexer.go]: done.

File: sysl2/lexer.go [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+		if err != nil {
+			panic("could not compile regex: " + reg)
+		}
+		compiledRegexes[i] = compiledRegex
+	}
+	ws, err := regexp.Compile(`[^ \t\r\n]+`)
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/lexer.go]: done.

File: sysl2/lexer.go [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+	if l.currentIndex < len(l.content) {
+		longestMatchLength := 0
+		locWS := l.WS.FindStringIndex(l.content[l.currentIndex:])
+		if locWS != nil {
+			start := l.currentIndex + locWS[0]
+			end := l.currentIndex + locWS[1]
+			str := l.content[start:end]
+
+			for i, r := range l.regexs {
+				loc := r.FindStringIndex(str)
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/lexer.go]: the WS has been renamed to skipWhitespaceRegex. and it does not need \A.

File: sysl2/parser.go (hunk 1 above)
Comment (sahejsingh) [reply sysl2/parser.go]: done.

File: sysl2/parser.go (hunk 10 above)
Comment (sahejsingh) [reply sysl2/parser.go]: done

File: sysl2/sysl_parser.go (hunk 2 above)
Comment (sahejsingh) [reply sysl2/sysl_parser.go]: removed the curly brackets.

File: sysl2/sysl_parser.go (hunk 3 above)
Comment (sahejsingh) [reply sysl2/sysl_parser.go]: done.

**Comment by sahejsingh** [comment]
Will open another for master.

**Pull request closed without merging**